### PR TITLE
EVSE add Session and Event support, test event trigger support

### LINKED ISSUE #30997: [1.3] Complete EVSE delegate implementation
### Reproduction steps

This issue is to track the outstanding work in the EVSE Delegate

Implemented in PR: https://github.com/project-chip/connectedhomeip/pull/30999/
- [x] - [x] Add Callback to HW to inform it of MaxCurrentLimit
- [x] Add Callback to HW to inform it of Change in Charging Enabled / Disable state

Implemented in PR: https://github.com/project-chip/connectedhomeip/pull/31200/
- [x] Add PluggedIn / Unplugged Events
- [x] Add EnergyTransferStarted Event
- [x] Add EnergyTransferStopped Event
- [x] Add Session support (Requires EEM cluster integration)
- [x] Add Persisted Storage support for some attributes

- [x] Add Timer expiry checking to automatically stop charging/discharging

Implemented in PR 
https://github.com/project-chip/connectedhomeip/pull/31406

-[x] Update XML for EVSE to support new charging preference structures

To do:
- [ ] Implement Persisting of helper data which allows session duration, session energy charged etc to be computed

- [ ] Add Charging Preferences Support (GetTargets/SetTargets/ClearTargets) (see PR https://github.com/project-chip/connectedhomeip/pull/31407)
- [ ] Add RPC handlers into example energy management app

### Bug prevalence

N/A

### GitHub hash of the SDK that was being used

N/A

### Platform

core

### Platform Version(s)

1.3

### Type

Core SDK Performance Improvement

### Anything else?

This is an issue to track EVSE implementation for 1.3

### LINKED ISSUE #31034: Add EVSE Python Test Scripts
There are no test scripts yet for EVSE cluster.

These should match with the test plan see:
https://github.com/CHIP-Specifications/chip-test-plans/pull/3709

We have a draft version to put up for PR.

### LINKED ISSUE #31061: Use correct Device Type ID for Energy Management App EP1 (EVSE)
EVSE in Example energy management app does not use correct device type ID. Device type file does not yet exist.

Edit: This affects `examples/energy-management-app/linux/include/CHIPProjectAppConfig.h` and the device type in the ZAP config.

### PR DESCRIPTION
**Pull request by jamesharrow**
Finishing off EVSE Cluster implementation:

Fixes #31034
 - (https://github.com/project-chip/connectedhomeip/issues/31034)

Fixes #31061 
 - (https://github.com/project-chip/connectedhomeip/issues/31061)

Relates to #30997 

- [x] Implement Session class to handle session info
- [x] Implement EVConnected Event
- [x] Implement  EVNotDetecteded Event
- [x] Implement  EnergyTransferStarted Event
- [x] Implement EnergyTransferStopped Event
- [x] Implement Fault Event support
- [x] Implement TestEventTrigger support to enable testing
- [x] Implement TC_EEVSE_2_2 to test most of the above - one aspect not yet working is the Timer checking (due to missing implementation to check ChargingEnabledUntil time).
- [x] Implement TC_EEVSE_2_4 to test Faults (needs an update to the TC_EEVSE test plan to add this) - so consider this experimental
- [x] Implement TC_EEVSE_2_5 to test Diagnostics (again not in official TC_EEVSE test plan yet)
- [x] Implement the Timeout (Charging Enabled Until)

### CODE REVIEW

File: src/app/clusters/energy-evse-server/EnergyEvseTestEventTriggerDelegate.h [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ */
+
+#pragma once
+
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/TestEventTriggerDelegate.h>
+
+namespace chip {
+
+enum class EnergyEvseTrigger : uint64_t
```
Comment (bzbarsky-apple) [inline src/app/clusters/energy-evse-server/EnergyEvseTestEventTriggerDelegate.h:25]: Should probably document how this corresponds to the delegate's `enableKey`.  I assume these 8-byte values are possible enable key values, but endianness needs to be documented or something....

Alternately, the constructor could take these values directly and do the endianness bits itself as needed, if that's the intent?

File: src/app/clusters/energy-evse-server/EnergyEvseTestEventTriggerDelegate.h [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    kEVPluggedInClear    = 0x0099000000000003,     // EV Plugged-in Test Event Clear | Simulate unplugging the EV
+    kEVChargeDemand      = 0x0099000000000004, // EV Charge Demand Test Event | Simulate the EV presenting charge demand to the EVSE
+    kEVChargeDemandClear = 0x0099000000000005, // EV Charge Demand Test Event Clear | Simulate the EV becoming fully charged
+
+};
+
+class EnergyEvseTestEventTriggerDelegate : public TestEventTriggerDelegate
+{
+public:
+    explicit EnergyEvseTestEventTriggerDelegate(const ByteSpan & enableKey, TestEventTriggerDelegate * otherDelegate) :
```
Comment (bzbarsky-apple) [inline src/app/clusters/energy-evse-server/EnergyEvseTestEventTriggerDelegate.h:42]: So this relies on the caller keeping the buffer `enableKey` points to alive as long as this object is alive, right?  At minimum, this needs to be documented.  But could also just make a local copy into a member instead...

Comment (jamesharrow) [reply src/app/clusters/energy-evse-server/EnergyEvseTestEventTriggerDelegate.h:42]: Again just copied the pattern from SmokeCOTest and OTATestEvent Trigger models.
So the same comment here would apply to them as well.

File: src/app/clusters/energy-evse-server/EnergyEvseTestEventTriggerDelegate.h (hunk 1 above)
Comment (jamesharrow) [reply src/app/clusters/energy-evse-server/EnergyEvseTestEventTriggerDelegate.h:25]: Ok - I was just following the same pattern as src/app/clusters/smoke-co-alarm-server/SmokeCOTestEventTriggerDelegate.h

These are the codes use to specific a specific type of test event trigger, not the enableKey.

Comment (jamesharrow) [reply src/app/clusters/energy-evse-server/EnergyEvseTestEventTriggerDelegate.h:32]: Added comment

Comment (bzbarsky-apple) [reply src/app/clusters/energy-evse-server/EnergyEvseTestEventTriggerDelegate.h:32]: Thank you, that makes a lot more sense.

File: src/app/clusters/energy-evse-server/EnergyEvseTestEventTriggerDelegate.h (hunk 2 above)
Comment (bzbarsky-apple) [reply src/app/clusters/energy-evse-server/EnergyEvseTestEventTriggerDelegate.h:48]: The other code may well be broken too....

Comment (jamesharrow) [reply src/app/clusters/energy-evse-server/EnergyEvseTestEventTriggerDelegate.h:48]: In terms of implementation not sure if you've found a security hole here?
When you run the application like example/energy-management app on linux, you need to add --enable-key arg with the magic secret key.

In https://github.com/project-chip/connectedhomeip/blob/a50f95964f2162f969a5e4e4f0b49a5ea1f0d998/examples/platform/linux/AppMain.cpp#L324-L369 there is a SampleTestEventTriggerDelegate class.

A bit later these other test event handlers are compiled into the application and each handler (one per cluster) gets added to a chain to be checked to see if it handles the Trigger, but in Linux at least I think it is handled by the SampleTestEventTrigger doing the checking of the enable key. However if that sample wasn't there then this implementation (like src/app/clusters/smoke-co-alarm-server/SmokeCOTestEventTriggerDelegate.cpp and src/app/clusters/ota-requestor/OTATestEventTriggerDelegate.cpp) doesn't have an Init().

See later in the same file:
https://github.com/project-chip/connectedhomeip/blob/a50f95964f2162f969a5e4e4f0b49a5ea1f0d998/examples/platform/linux/AppMain.cpp#L542-L560

Other platforms seem to have a static enableKey which may come from the factory data provider:
e.g. 
https://github.com/project-chip/connectedhomeip/blob/a50f95964f2162f969a5e4e4f0b49a5ea1f0d998/examples/light-switch-app/infineon/cyw30739/src/main.cpp#L64-L66

Comment (bzbarsky-apple) [reply src/app/clusters/energy-evse-server/EnergyEvseTestEventTriggerDelegate.h:61]: My point is that nothing in the API makes it clear that the thing passed in needs to have "lives forever" lifetime.  Yes, current simple things we have use "forever" lifetimes because it's convenient, but nothing prevents someone writing an app from loading this thing from flash and passing it in... and then dropping the buffer.  And nothing documents they shouldn't do that.

Comment (jamesharrow) [reply src/app/clusters/energy-evse-server/EnergyEvseTestEventTriggerDelegate.h:61]: I've added a comment to the header to help document this point.

File: src/python_testing/TC_EEVSE_Utils.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+                                             expected_status: Status = Status.Success):
+        try:
+            await self.send_single_cmd(cmd=Clusters.EnergyEvse.Commands.StartDiagnostics(),
+                                       endpoint=endpoint,
+                                       timedRequestTimeoutMs=timedRequestTimeoutMs)
+
+        except InteractionModelError as e:
+            asserts.assert_equal(e.status, expected_status, "Unexpected error returned")
+
+    async def send_test_event_triggers(self, enableKey=bytes([b for b in range(16)]), eventTrigger=0x0099000000000000):
```
Comment (cecille) [inline src/python_testing/TC_EEVSE_Utils.py:105]: Is this a hardcoded enable key?

Comment (jamesharrow) [reply src/python_testing/TC_EEVSE_Utils.py:105]: It is currently, I was looking for a way in matter_testing_support to be able to get this value but for now it is.

is there something in the framework to get this?

similarly the endpoint id is hardcoded as 1 for now.

Comment (jamesharrow) [reply src/python_testing/TC_EEVSE_Utils.py:105]: Is config.global_test_params the right place for it? Or should this come from a PICS file? (or both)?

It could come from a `--hex-arg enableKey==000102030405060708090A0B0C0D0E0F` ?

I managed to get this to work, but not sure if we should add something more specific to matter_testing_support.py as an explicit argument?


```    async def send_test_event_triggers(self, enableKey: bytes = None, eventTrigger=0x0099000000000000):
        # get the test event enable key or assume the default
        # This can be passed in on command line using 
        #    --hex-arg enableKey:000102030405060708090a0b0c0d0e0f 
        if enableKey is None:
            if 'enableKey' not in self.matter_test_config.global_test_params:
                enableKey = bytes([b for b in range(16)])
            else:
                enableKey = self.matter_test_config.global_test_params['enableKey'])

        try:
            await self.send_single_cmd(endpoint=0,
                                       cmd=Clusters.GeneralDiagnostics.Commands.TestEventTrigger(
                                           enableKey,
                                           eventTrigger)
                                       )

        except InteractionModelError as e:
            asserts.fail(f"Unexpected error returned - {e.status}")
         
```

File: examples/energy-management-app/energy-management-common/include/EVSEManufacturerImpl.h [hunk 4]
```diff
@@ -33,22 +33,37 @@ namespace EnergyEvse {
...
     /**
      * @brief   Main Callback handler from delegate to user code
      */
     static void ApplicationCallbackHandler(const EVSECbInfo * cb, intptr_t arg);
 
 private:
+    EnergyEvseManager * mInstance;
+
+    int64_t mLastChargingEnergyMeter    = 0;
+    int64_t mLastDischargingEnergyMeter = 0;
```
[the remark this answers is not in the document]
Comment (jamesharrow) [reply examples/energy-management-app/energy-management-common/include/EVSEManufacturerImpl.h:66]: In Matter all of the energy units are mWh - and this is why we had to use int64_t everywhere.
Maybe we could create a new typedef (like percent, epoch_s).

When we did the ZAP changes to add the autogen from the XML it just converted them to int64_t.
I wasn't sure how to add a new Typedef like energy_mwh types (such as percent).

Slightly reluctant to rename all the variables since we've agreed that these will always be in mW, mV, mWh, mA.
and I've tried to be consistent with the attribute names (based on the names in the specs - which don't include the units).

Whilst these in this case aren't attributes they are just copies of the attribute values.
If it does help I can change the variable names here.

File: examples/energy-management-app/energy-management-common/include/EnergyEvseDelegateImpl.h [hunk 5]
```diff
@@ -31,6 +31,68 @@ namespace app {
...
+{
+public:
+    EvseSession(EndpointId aEndpoint) { mEndpointId = aEndpoint; }
+    /**
+     * @brief This function samples the start-time, and energy meter to hold the session info
+     *
+     * @param chargingMeterValue    - The current value of the energy meter (charging) in mWh
+     * @param dischargingMeterValue - The current value of the energy meter (discharging) in mWh
+     */
+    void StartSession(int64_t chargingMeterValue, int64_t dischargingMeterValue);
```
[the remark this answers is not in the document]
Comment (jamesharrow) [reply examples/energy-management-app/energy-management-common/include/EnergyEvseDelegateImpl.h:61]: See my response above to why I'm reluctanct to do that.
All energy values in Matter are mWh.

File: examples/energy-management-app/energy-management-common/include/EnergyEvseDelegateImpl.h [hunk 6]
```diff
@@ -143,10 +223,11 @@ class EnergyEvseDelegate : public EnergyEvse::Delegate
 
 private:
     /* Constants */
-    static constexpr int DEFAULT_MIN_CHARGE_CURRENT          = 6000;                  /* 6A */
-    static constexpr int DEFAULT_USER_MAXIMUM_CHARGE_CURRENT = kMaximumChargeCurrent; /* 80A */
-    static constexpr int DEFAULT_RANDOMIZATION_DELAY_WINDOW  = 600;                   /* 600s */
-    static constexpr int kMaxVehicleIDBufSize                = 32;
+    static constexpr int DEFAULT_MIN_CHARGE_CURRENT                         = 6000;                  /* 6A */
```
[the remark this answers is not in the document]
Comment (jamesharrow) [reply examples/energy-management-app/energy-management-common/include/EnergyEvseDelegateImpl.h:226]: There seems to be a mix of conventions used - but will update. Thanks

File: examples/energy-management-app/energy-management-common/src/EnergyEvseManager.cpp [hunk 7]
```diff
@@ -17,14 +17,101 @@
...
+        ChipLogError(AppServer, "EVSE: Unable to restore persisted ApproximateEVEfficiency value");
+    }
+
+    return CHIP_NO_ERROR; // It is ok to have no value loaded here
+}
+
 CHIP_ERROR EnergyEvseManager::Init()
 {
-    return Instance::Init();
+    CHIP_ERROR err = Instance::Init();
```
[the remark this answers is not in the document]
Comment (jamesharrow) [reply examples/energy-management-app/energy-management-common/src/EnergyEvseManager.cpp:108]: Done

File: examples/energy-management-app/energy-management-common/src/EnergyEvseDelegateImpl.cpp [hunk 8]
```diff
@@ -831,17 +1420,142 @@ DataModel::Nullable<CharSpan> EnergyEvseDelegate::GetVehicleID()
...
+ * @param chargingMeterValue    - The current value of the energy meter (charging) in mWh
+ * @param dischargingMeterValue - The current value of the energy meter (discharging) in mWh
+ */
+void EvseSession::StartSession(int64_t chargingMeterValue, int64_t dischargingMeterValue)
+{
+    /* Get Timestamp */
+    uint32_t chipEpoch = 0;
+    if (GetEpochTS(chipEpoch) != CHIP_NO_ERROR)
+    {
+        ChipLogError(AppServer, "EVSE: Failed to get timestamp when starting session");
```
[the remark this answers is not in the document]
Comment (jamesharrow) [reply examples/energy-management-app/energy-management-common/src/EnergyEvseDelegateImpl.cpp:1486]: There is a couple of ChipLogErrors in GetEpochTS()
See Line 1460 and 1467?

```CHIP_ERROR GetEpochTS(uint32_t & chipEpoch)
{
    chipEpoch = 0;

    System::Clock::Milliseconds64 cTMs;
    CHIP_ERROR err = System::SystemClock().GetClock_RealTimeMS(cTMs);

    /* If the GetClock_RealTimeMS returns CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE, then
     * This platform cannot ever report real time !
     * This should not be certifiable since getting time is a Mandatory
     * feature of EVSE Cluster
     */
    VerifyOrDie(err != CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE);

    if (err != CHIP_NO_ERROR)
    {
        ChipLogError(Zcl, "EVSE: Unable to get current time - err:%" CHIP_ERROR_FORMAT, err.Format());
        return err;
    }

    auto unixEpoch = std::chrono::duration_cast<System::Clock::Seconds32>(cTMs).count();
    if (!UnixEpochToChipEpochTime(unixEpoch, chipEpoch))
    {
        ChipLogError(Zcl, "EVSE: unable to convert Unix Epoch time to Matter Epoch Time");
        return err;
    }

    return CHIP_NO_ERROR;
}
```
Or did you mean somewhere else?

File: examples/energy-management-app/energy-management-common/src/EnergyEvseDelegateImpl.cpp [hunk 9]
```diff
@@ -306,39 +324,69 @@ Status EnergyEvseDelegate::HwSetCableAssemblyLimit(int64_t currentmA)
...
+            mHwState = newState;
+            HandleStateMachineEvent(EVSEStateMachineEvent::EVDemandEvent);
+        }
         break;
 
     default:
         /* All other states should be managed by the Delegate */
-        // TODO (assert?)
+        ChipLogError(AppServer, "HwSetState received invalid enum from caller");
+        return Status::Failure;
```
[the remark this answers is not in the document]
Comment (jamesharrow) [reply examples/energy-management-app/energy-management-common/src/EnergyEvseDelegateImpl.cpp:389]: Done

File: examples/energy-management-app/energy-management-common/src/EnergyEvseDelegateImpl.cpp [hunk 10]
```diff
@@ -306,39 +324,69 @@ Status EnergyEvseDelegate::HwSetCableAssemblyLimit(int64_t currentmA)
...
-Status EnergyEvseDelegate::HwSetState(StateEnum state)
+Status EnergyEvseDelegate::HwSetState(StateEnum newState)
 {
-    switch (state)
+    switch (newState)
     {
     case StateEnum::kNotPluggedIn:
-        // TODO - work out logic here
-        mHwState = state;
+        if (mHwState == StateEnum::kPluggedInNoDemand || mHwState == StateEnum::kPluggedInDemand)
```
[the remark this answers is not in the document]
Comment (jamesharrow) [reply examples/energy-management-app/energy-management-common/src/EnergyEvseDelegateImpl.cpp:348]: all of the else's are deliberately nothing of importance happens.

I did consider having nested switch statements but felt that someone would complain about that being unwieldy and difficult to read & follow.

I'll see if I can go for the switch with a sub switch.

File: examples/energy-management-app/energy-management-common/src/EnergyEvseDelegateImpl.cpp
Comment (bzbarsky-apple) [inline examples/energy-management-app/energy-management-common/src/EnergyEvseDelegateImpl.cpp:1]: I did not review this file carefully.  Please make sure someone does.

File: src/app/clusters/energy-evse-server/EnergyEvseTestEventTriggerDelegate.h [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+        mEnableKey(enableKey), mOtherDelegate(otherDelegate)
+    {}
+
+    /**
+     * This function expects the enableKey ByteSpan to be valid forever.
```
Comment (bzbarsky-apple) [inline src/app/clusters/energy-evse-server/EnergyEvseTestEventTriggerDelegate.h:66]: No, this function expects nothing of the sort.

It's the _constructor_ that expects that, no?  Specifically that the data pointed to by the ByteSpan will never change or disappear.

File: src/python_testing/TC_EEVSE_2_2.py
Comment (bzbarsky-apple) [inline src/python_testing/TC_EEVSE_2_2.py:1]: I did not review the tests carefully either.  Please make sure someone does.

File: src/python_testing/TC_EEVSE_Utils.py (hunk 3 above)
Comment (jamesharrow) [reply src/python_testing/TC_EEVSE_Utils.py:105]: Based on slack conversation I'll commit this as the way to go.

File: examples/energy-management-app/energy-management-common/include/EnergyEvseDelegateImpl.h [hunk 12]
```diff
@@ -74,15 +136,33 @@ class EnergyEvseDelegate : public EnergyEvse::Delegate
      */
     Status HwRegisterEvseCallbackHandler(EVSECallbackFunc handler, intptr_t arg);
 
+    /**
+     * @brief   This is used to start a check on if the Enabled timer needs to be started
```
[the remark this answers is not in the document]
Comment (jamesharrow) [reply examples/energy-management-app/energy-management-common/include/EnergyEvseDelegateImpl.h:140]: Will rephase... 
Meant to say "Start a timer to check..." (was late writing this!)

File: src/python_testing/TC_EEVSE_2_2.py
Comment (jamesharrow) [reply src/python_testing/TC_EEVSE_2_2.py:1]: I've had some internal reviews pick up issues against the specs.

File: src/app/clusters/energy-evse-server/EnergyEvseTestEventTriggerDelegate.h (hunk 11 above)
Comment (jamesharrow) [reply src/app/clusters/energy-evse-server/EnergyEvseTestEventTriggerDelegate.h:66]: Yep - was late.

File: examples/energy-management-app/energy-management-common/src/EnergyEvseDelegateImpl.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (jamesharrow) [reply examples/energy-management-app/energy-management-common/src/EnergyEvseDelegateImpl.cpp]: Made changes. Please re-review.

File: examples/energy-management-app/energy-management-common/include/EVSECallbacks.h [hunk 13]
```diff
@@ -44,12 +44,22 @@ enum EVSECallbackType
      * Charging Preferences have changed
      */
     ChargingPreferencesChanged,
+    /*
+     * Energy Meter Reading requested
```
[the remark this answers is not in the document]
Comment (jamesharrow) [reply examples/energy-management-app/energy-management-common/include/EVSECallbacks.h]: Added more descriptions.

File: examples/energy-management-app/energy-management-common/include/EVSECallbacks.h [hunk 14]
```diff
@@ -44,12 +44,22 @@ enum EVSECallbackType
...
+     */
+    EnergyMeterReadingRequested,
     /*
      * DeviceEnergyManagement has changed
      */
     DeviceEnergyManagementChanged,
 };
 
+enum ChargingDischargingType
```
[the remark this answers is not in the document]
Comment (jamesharrow) [reply examples/energy-management-app/energy-management-common/include/EVSECallbacks.h]: Done

File: examples/energy-management-app/energy-management-common/src/EnergyEvseDelegateImpl.cpp
Comment (jamesharrow) [reply examples/energy-management-app/energy-management-common/src/EnergyEvseDelegateImpl.cpp:1]: Have had internal review already but would welcome others.

File: examples/energy-management-app/energy-management-common/src/EnergyEvseDelegateImpl.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (andy31415) [reply examples/energy-management-app/energy-management-common/src/EnergyEvseDelegateImpl.cpp:1515]: Did not see that they were logged below. My general expectation is that anything that receives errors has 2 choices:
   - report then up the stack
   - drop them to the floor, but in this case log them
 
I saw your code did the drop to the floor but no logging. I am worried that if you rely that lower levels always log before returning any error then this will not be very maintainable. I would actually consider logging twice in this case. 

You could also document that `GetEpochTS(chipEpoch)`  will *always* log any returned errors so that callers do not need to log, however that seems slightly awkward to me.

File: examples/energy-management-app/energy-management-common/src/EVSEManufacturerImpl.cpp [hunk 15]
```diff
@@ -18,53 +18,57 @@
...
 #include <EnergyEvseManager.h>
+#include <app/clusters/energy-evse-server/EnergyEvseTestEventTriggerDelegate.h>
 
+using namespace chip;
 using namespace chip::app;
 using namespace chip::app::Clusters;
 using namespace chip::app::Clusters::EnergyEvse;
 
-CHIP_ERROR EVSEManufacturer::Init(EnergyEvseManager * aInstance)
+/* Function prototype - this should be implemented somewhere in main.cpp or similar */
```
[the remark this answers is not in the document]
Comment (andy31415) [reply examples/energy-management-app/energy-management-common/src/EVSEManufacturerImpl.cpp:28]: Please place this in a separate header to make it clear about it being needed from somewhere else.

File: examples/energy-management-app/energy-management-common/src/EnergyEvseDelegateImpl.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (andy31415) [reply examples/energy-management-app/energy-management-common/src/EnergyEvseDelegateImpl.cpp:1515]: maybe add a comment that reason is logged by lower levels. at the same time I believe just double-logging the error code is probably ok too ... won't cause harm and better to be sure this can be debugged.

[the remark this answers is not in the document]
Comment (jamesharrow) [reply examples/energy-management-app/energy-management-common/src/EnergyEvseDelegateImpl.cpp:1515]: Added improved error logging in both places and improved comments.

File: examples/energy-management-app/energy-management-common/src/EVSEManufacturerImpl.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (jamesharrow) [reply examples/energy-management-app/energy-management-common/src/EVSEManufacturerImpl.cpp]: Moved to EVSEManufacturerImpl.h and added comment.

**Review by andy31415 — APPROVED** [review]

**Review by bzbarsky-apple — APPROVED** [review]

File: src/python_testing/TC_EEVSE_Utils.py [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+import chip.clusters as Clusters
+from chip.clusters import ClusterObjects as ClusterObjects
+from chip.clusters.Attribute import EventReadResult, SubscriptionTransaction
+from chip.interaction_model import InteractionModelError, Status
+from mobly import asserts
+
+logger = logging.getLogger(__name__)
+
+
+class EventChangeCallback:
```
Comment (cecille) [inline src/python_testing/TC_EEVSE_Utils.py:30]: This looks generally quite helpful. Don't block this PR on it, but would you mind moving this into the base helper file so other folks could use it? It doesn't appear to be specific to this cluster

File: src/python_testing/TC_EEVSE_Utils.py [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+    async def write_user_max_charge(self, endpoint: int = None, user_max_charge: int = 0):
+        if endpoint is None:
+            endpoint = self.matter_test_config.endpoint
+        result = await self.default_controller.WriteAttribute(self.dut_node_id,
+                                                              [(endpoint,
+                                                               Clusters.EnergyEvse.Attributes.UserMaximumChargeCurrent(user_max_charge))])
+        asserts.assert_equal(result[0].Status, Status.Success, "UserMaximumChargeCurrent write failed")
+
+    async def send_enable_charge_command(self, endpoint: int = None, charge_until: int = None, timedRequestTimeoutMs: int = 3000,
+                                         min_charge: int = None, max_charge: int = None, expected_status: Status = Status.Success):
```
Comment (cecille) [inline src/python_testing/TC_EEVSE_Utils.py:82]: min charge and max charge aren't nullable, so setting the default to none here may cause weird errors. Is there a good default? 0?

File: src/python_testing/TC_EEVSE_2_2.py [hunk 18]
```diff
@@ -0,0 +1,319 @@
...
+        self.step("4a")
+        await self.check_evse_attribute("State", Clusters.EnergyEvse.Enums.StateEnum.kPluggedInNoDemand)
+
+        self.step("4b")
+        # Save Session ID for later and check it against the value in the event
+        session_id = await self.read_evse_attribute_expect_success(attribute="SessionID")
+        self.validate_ev_connected_event(event_data, session_id)
+
+        self.step("5")
+        charging_duration = 5  # TODO test plan spec says 120s - reduced for now
```
Comment (cecille) [inline src/python_testing/TC_EEVSE_2_2.py:137]: Is this still a reasonable test? ie, is the intent here to change the test plan, or was this reduced for development and not changed back?

Note that we do have a PICS for CI so if you want to differentiate the CI from the cert test, that's also possible.

File: src/python_testing/TC_EEVSE_2_2.py [hunk 19]
```diff
@@ -0,0 +1,319 @@
...
+        # from step 8 above - validate event
+        self.validate_energy_transfer_started_event(event_data, session_id, expected_state, expected_max_charge)
+
+        self.step("9")
+        # This will only work if the optional UserMaximumChargeCurrent attribute is supported
+        supported_attributes = await self.get_supported_energy_evse_attributes()
+        if Clusters.EnergyEvse.Attributes.UserMaximumChargeCurrent.attribute_id in supported_attributes:
+            logging.info("UserMaximumChargeCurrent is supported...")
+            user_max_charge_current = 6000
+            await self.write_user_max_charge(1, user_max_charge_current)
```
Comment (cecille) [inline src/python_testing/TC_EEVSE_2_2.py:219]: There's a verification step in the test plan here, but I'm not sure what it means - it says "Charging rate reduced to 6A". Is that step 6a? Do you need to check that? Seems like it might be checked below. Is that just test plan weirdness?

**Review by cecille — COMMENTED** [review]
I can't find the test plans for 2.4, 2.5, but looked over 2.2, and it seems reasonable. The rest also look reasonable, but I wasn't able to compare to the test plans, so who knows.

You have a check from google already, so not going to approve since I didn't look over any of the non-test stuff.

File: src/python_testing/TC_EEVSE_2_2.py (hunk 18 above)
Comment (jamesharrow) [reply src/python_testing/TC_EEVSE_2_2.py:137]: I think we should probably change the test plan to a smaller number too, originally this 2 mins was chosen because the implementation around timers wasn't certain (e.g. the EVSE might run a 1 min timer check, but now it is implemented it can be a much smaller number). 

Its not currently in the CI checkers either - I'll raise a follow up on both these aspects and to change this value to a suitable number.

File: src/python_testing/TC_EEVSE_Utils.py (hunk 17 above)
Comment (jamesharrow) [reply src/python_testing/TC_EEVSE_Utils.py:82]: I think I left it as None so it would assert if these values weren't supplied, rather than sending 0 which may be treated as valid values and then potentially obscure an implementation error in the DUT if the test isn't good enough at checking the charging current used properly.

I'll refactor to ensure that they are always supplied by adding a test in the function to check it isn't None (rather than having to reorder the args which python insists on).

File: src/python_testing/TC_EEVSE_2_2.py (hunk 19 above)
Comment (jamesharrow) [reply src/python_testing/TC_EEVSE_2_2.py:219]: That check is done on line 224-225 below.
It is 6Amps. but the algorithm is based on the expected_max_charge = min(user_max_charge_current, circuit_capacity)

File: src/python_testing/TC_EEVSE_2_2.py (hunk 18 above)
Comment (jamesharrow) [reply src/python_testing/TC_EEVSE_2_2.py:137]: Added Follow Issue https://github.com/project-chip/connectedhomeip/issues/31460

File: src/python_testing/TC_EEVSE_Utils.py (hunk 17 above)
Comment (jamesharrow) [reply src/python_testing/TC_EEVSE_Utils.py:82]: Added follow up PR https://github.com/project-chip/connectedhomeip/issues/31460

File: src/python_testing/TC_EEVSE_Utils.py (hunk 16 above)
Comment (jamesharrow) [reply src/python_testing/TC_EEVSE_Utils.py:30]: I've add a follow Issue https://github.com/project-chip/connectedhomeip/issues/31460

**Review by michaelleecowan — APPROVED** [review]
Approve, looks good
